File: 144-BinaryTreePreorderTraversal/144-BinaryTreePreorderTraversal.py
# Last updated: 4/21/2026, 11:26:49 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
   def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.tr = []  
        self._preorder_helper(root) 
        return self.tr

   def _preorder_helper(self, node: Optional[TreeNode]):
        if node is None:
            return
        self.tr.append(node.val)  
        self._preorder_helper(node.left)  
        self._preorder_helper(node.right)  
        
        

        

File: 144-BinaryTreePreorderTraversal/144-BinaryTreePreorderTraversal.cpp
// Last updated: 4/21/2026, 11:26:42 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 #include <vector>
 #include <stack>
 #include <iostream>
 using namespace std;
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> preorder;
        stack<TreeNode*> st;
        if(root == NULL){
            return preorder;
        }
        st.push(root);
        while(!st.empty()){
            TreeNode* curr = st.top();
            st.pop();
            preorder.push_back(curr->val);
            if(curr->right){
                st.push(curr->right);
            }
            if(curr->left){
                st.push(curr->left);
            }
        }
        return preorder;
        
    }
};